packit_service/worker/build/copr_build.py

```diff
@@ -185,7 +192,137 @@ def monitor_not_submitted_copr_builds(self, number_of_builds: int, reason: str):
+
+        Returns: list of urls
+        """
+        try:
+            latest_successful_build_id = (
+                self.api.copr_helper.copr_client.package_proxy.get(
+                    "packit",
+                    "packit-stable",
+                    "packit",
+                    with_latest_succeeded_build=True,
+                ).builds["latest_succeeded"]["id"]
+            )
+            result_url = self.api.copr_helper.copr_client.build_chroot_proxy.get(
+                latest_successful_build_id, "fedora-35-x86_64"
+            ).result_url
+            package_nvrs = self.get_built_packages(
+                latest_successful_build_id, "fedora-35-x86_64"
+            )
+            built_packages = get_package_nvrs(package_nvrs)
+            return [f"{result_url}{package}.rpm" for package in built_packages]
+        except Exception as ex:
+            logger.debug(
+                f"Getting download urls for latest packit/packit-stable "
+                f"build was not successful: {ex}"
+            )
+            raise ex
+
+    def run_copr_build_from_source_script(self) -> TaskResults:
+        """
+        Run copr build using custom source method.
+        """
+        try:
+            build_id, web_url = self.submit_copr_build(from_source_script=True)
+        except Exception as ex:
+            self.handle_build_submit_error(ex)
+            return TaskResults(
+                success=False,
+                details={"msg": "Submit of the Copr build failed.", "error": str(ex)},
+            )
+
+        (self._srpm_model, self.run_model,) = SRPMBuildModel.create_with_new_run(
+            copr_build_id=str(build_id),
+            commit_sha=self.metadata.commit_sha,
+            trigger_model=self.db_trigger,
+            copr_web_url=web_url,
+        )
+
+        self.report_status_to_all(
+            description="SRPM build in Copr was submitted...",
+            state=BaseCommitStatus.pending,
+            url=get_srpm_build_info_url(self.srpm_model.id),
+        )
+
+        self.handle_rpm_build_start(build_id, web_url, waiting_for_srpm=True)
+
+        return TaskResults(success=True, details={})
+
+    def submit_copr_build(self, from_source_script: bool = False) -> Tuple[int, str]:
+        """
+        Create the project in Copr if not exists and submit a new build using
+        source script method
+        Return:
+            tuple of build ID and web url
+        """
+        owner = self.create_copr_project_if_not_exists()
+        try:
+            if from_source_script:
+                script = create_source_script(
+                    url=f"{self.project.service.hostname}/{self.project.full_repo_name}",
+                    ref=self.metadata.git_ref,
+                    pr_id=str(self.metadata.pr_id),
+                    merge_pr=self.package_config.merge_pr_in_ci,
+                    job_config=self.job_config,
+                )
```

Comment:
What do you think about sending the script as an argument? (Just an idea.)

Reply:
Sounds good, I will try it that way.